clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -534,30 +524,29 @@ private void handleCompletedReceives(List<ClientResponse> responses, long now) {
             if (req.isInternalRequest && body instanceof MetadataResponse)
                 metadataUpdater.handleCompletedMetadataResponse(req.header, now, (MetadataResponse) body);
             else if (req.isInternalRequest && body instanceof ApiVersionsResponse)
-                handleApiVersionsResponse(req, (ApiVersionsResponse) body);
+                handleApiVersionsResponse(responses, req, now, (ApiVersionsResponse) body);
             else
                 responses.add(req.completed(body, now));
         }
     }
 
-    private void handleApiVersionsResponse(InFlightRequest req, ApiVersionsResponse apiVersionsResponse) {
+    private void handleApiVersionsResponse(List<ClientResponse> responses,
+            InFlightRequest req, long now, ApiVersionsResponse apiVersionsResponse) {
         final String node = req.destination;
-        for (ApiVersionsResponse.ApiVersion requiredApiVersion : requiredApiVersions) {
-            final ApiVersionsResponse.ApiVersion supportedApiVersion = apiVersionsResponse.apiVersion(requiredApiVersion.apiKey);
-            if (supportedApiVersion == null) {
-                close(node);
-                throw new KafkaException("Node " + req.destination + " does not support Api " + requiredApiVersion.apiKey + ".");
-            }
-            if (supportedApiVersion.maxVersion < requiredApiVersion.minVersion || supportedApiVersion.minVersion > requiredApiVersion.maxVersion) {
-                close(node);
-                throw new KafkaException("Node " + req.destination + " does not support required versions for Api " + requiredApiVersion.apiKey + "." +
-                        " Supported versions: " + "[" + supportedApiVersion.minVersion + ", " + supportedApiVersion.maxVersion + "]" + "," +
-                        " required versions: " + "[" + requiredApiVersion.minVersion + ", " + requiredApiVersion.maxVersion + "]" + ".");
-            }
+        if (apiVersionsResponse.errorCode() != Errors.NONE.code()) {
+            log.warn("Node " + node + " got error " + apiVersionsResponse.errorCode() +
+                    " when making an ApiVersionsRequest.  Disconnecting.");
+            this.selector.close(node);
+            processDisconnection(responses, node, now);
```

Comment:
yeah, that is pretty nasty.  I'll fix it.  I wish there were some workaround for this, but it looks like Java Map still always supports put(Object) to allow compatibility with the pre-generics days...

We should start running findBugs to catch things like this